### ISSUE TITLE
Array fields consisting of tuples fail to deserialise

### ISSUE THREAD

**Issue by romanzes** [kardeiz/serde_tuple#6]
```rust
#[derive(Serialize, Deserialize)]
struct Foo {
    bars: Vec<Bar>,
}

#[derive(Serialize_tuple, Deserialize_tuple)]
struct Bar {
    length: i32,
}

let foo = Foo {
    bars: vec![Bar { length: 1 }, Bar { length: 2 }]
};
let serialized = serde_json::to_string(&foo).unwrap();
serde_json::from_str::<Foo>(&serialized).expect("");

// or just

serde_json::from_str::<Foo>("{\"bars\":[[1],[2]]}").expect("");

// or even

serde_json::from_str::<Vec<Bar>>("[[1],[2]]").expect("");
```
Leads to `Error("invalid type: sequence, expected i32", line: 1, column: 1)`

**Comment by kardeiz** [comment]
@romanzes Good point, thanks for the issue!

This is because `serde` serializes `struct Wrapper(i32,)` as, for example, `1`, but `struct Wrapper(i32, i32)` as `[1, 2]`.

I think this is usually the right thing: many/most single-field structs are new-type wrappers, where the `inner`/`0` field should be de/serialized directly.

It would be possible to modify this library to use `Wrapper((i32,))` internally (that is, wrap the inner fields in a tuple), which would provide your expected behavior. But I'm not sure if this would be the best idea.

It would also be possible to add the ability to deserialize from either `1` or `[1]` pretty unobtrusively, but I think serialization has to make a choice. We could also add an attribute to specify whether to always use a sequence for single-field structs.

@dtolnay, would you mind taking a look at this issue and sharing your thoughts? Thanks!

**Comment by dtolnay** [comment]
Hi @kardeiz, `Wrapper((i32,))` wouldn't work in general because you could no longer preserve the field attributes, i.e. `Wrapper((#[serde(...)] i32,))` isn't legal. What you could do instead is:

- For serialize, instead of calling `serde::Serialize::serialize(&inner, serializer)`, wrap the `serializer` argument in your own Serializer impl whose `serialize_newtype_struct` method transforms into calling `serialize_tuple_struct` on the underlying serializer.

- Same thing for deserialize; wrap the `deserializer` argument in your own Deserializer impl whose `deserialize_newtype_struct` method calls `deserialize_tuple_struct` on the real deserializer.

**Comment by kardeiz** [comment]
@dtolnay, Thanks for the suggestions! Right, I forgot about field attributes. I'll work on getting this set up.

**Comment by kardeiz** [comment]
@romanzes This is now fixed in [`0.5.0`](https://crates.io/crates/serde_tuple/0.5.0) and should provide your desired behavior. 

Closing this, but please let me know if you have any further issues&mdash;
